### PR TITLE
fuzz core/blocks package

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #4882

---

# Description
Fuzz test state package and fix errors that have been caught
will include skip slot cache in another pr

### CODE REVIEW

File: beacon-chain/core/blocks/block_operations.go [hunk 1]
```diff
@@ -103,6 +103,15 @@ func verifySignature(signedData []byte, pub []byte, signature []byte, domain uin
 //    if state.eth1_data_votes.count(body.eth1_data) * 2 > SLOTS_PER_ETH1_VOTING_PERIOD:
 //        state.latest_eth1_data = body.eth1_data
 func ProcessEth1DataInBlock(beaconState *stateTrie.BeaconState, block *ethpb.BeaconBlock) (*stateTrie.BeaconState, error) {
+	if beaconState == nil {
+		return nil, errors.New("nil state")
+	}
+	if block == nil {
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: wouldn't it be more efficient just do block == nil and block.Body == nil in one line?

File: beacon-chain/core/blocks/block_operations.go [hunk 2]
```diff
@@ -742,7 +777,9 @@ func ProcessAttestationNoVerify(
 func VerifyIndexedAttestation(ctx context.Context, beaconState *stateTrie.BeaconState, indexedAtt *ethpb.IndexedAttestation) error {
 	ctx, span := trace.StartSpan(ctx, "core.VerifyIndexedAttestation")
 	defer span.End()
-
+	if indexedAtt == nil || indexedAtt.Data == nil || indexedAtt.Data.Target == nil {
+		return fmt.Errorf("nil or missing indexed attestation data: %v", indexedAtt)
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: why are we logging the indexedAtt if it's nil or missing?

Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: why are we logging the indexedAtt if it's nil or missing?

File: beacon-chain/core/blocks/block_operations.go [hunk 3]
```diff
@@ -804,6 +841,9 @@ func VerifyIndexedAttestation(ctx context.Context, beaconState *stateTrie.Beacon
 // VerifyAttestation converts and attestation into an indexed attestation and verifies
 // the signature in that attestation.
 func VerifyAttestation(ctx context.Context, beaconState *stateTrie.BeaconState, att *ethpb.Attestation) error {
+	if att == nil || att.Data == nil {
+		return fmt.Errorf("nil or missing indexed attestation data: %v", att)
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]:
```suggestion
		return fmt.Errorf("nil or missing attestation data: %v", att)
```

File: beacon-chain/core/blocks/block_operations.go [hunk 4]
```diff
@@ -830,6 +870,9 @@ func ProcessDeposits(
 	var err error
 	deposits := body.Deposits
 	for _, deposit := range deposits {
+		if deposit == nil || deposit.Data == nil {
+			return nil, fmt.Errorf("nil deposit in block body: %v", body)
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: shouldn't we just continue here?

File: beacon-chain/core/blocks/block_operations.go [hunk 5]
```diff
@@ -1024,6 +1074,9 @@ func ProcessVoluntaryExits(
 ) (*stateTrie.BeaconState, error) {
 	exits := body.VoluntaryExits
 	for idx, exit := range exits {
+		if exit == nil || exit.Exit == nil {
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: Just continue here?

File: beacon-chain/core/helpers/validators.go [hunk 6]
```diff
@@ -244,6 +244,9 @@ func ComputeProposerIndex(validators []*ethpb.Validator, activeIndices []uint64,
 //    fork_version = state.fork.previous_version if epoch < state.fork.epoch else state.fork.current_version
 //    return bls_domain(domain_type, fork_version)
 func Domain(fork *pb.Fork, epoch uint64, domainType []byte) uint64 {
+	if fork == nil || domainType == nil {
+		return 0
```
Comment (terencechain) [inline beacon-chain/core/helpers/validators.go]: 0 is a valid domain no? shouldn't we return a max int in this case?

File: beacon-chain/core/validators/validator.go [hunk 7]
```diff
@@ -38,6 +40,9 @@ import (
 //    validator.withdrawable_epoch = Epoch(validator.exit_epoch + MIN_VALIDATOR_WITHDRAWABILITY_DELAY)
 func InitiateValidatorExit(state *stateTrie.BeaconState, idx uint64) (*stateTrie.BeaconState, error) {
 	vals := state.Validators()
+	if idx >= uint64(len(vals)) {
+		return nil, fmt.Errorf("requested validator idx: %d to exit is higher then validator count: %d", idx, len(vals))
```
Comment (terencechain) [inline beacon-chain/core/validators/validator.go]:
```suggestion
		return nil, fmt.Errorf("validator idx %d is higher then validator count %d", idx, len(vals))
```

File: beacon-chain/state/getters.go [hunk 8]
```diff
@@ -13,36 +13,54 @@ import (
...
+		return 0
+	}
 	return v.validator.EffectiveBalance
 }
 
 // ActivationEligibilityEpoch returns the activation eligibility epoch of the
 // read only validator.
 func (v *ReadOnlyValidator) ActivationEligibilityEpoch() uint64 {
+	if v == nil || v.validator == nil {
+		return 0
```
Comment (terencechain) [inline beacon-chain/state/getters.go]: given 0 is a valid epoch but im sure if this is a sound approach, return max int?

File: beacon-chain/core/blocks/block_operations.go (hunk 1 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: one line now. thanks

File: beacon-chain/core/blocks/block_operations.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: removed
thanks

Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: duplicate

File: beacon-chain/core/blocks/block_operations.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: thanks

File: beacon-chain/core/blocks/block_operations.go (hunk 4 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: not sure about it
why should ssz deserialise deposits with a nil entry?
wdyt?

File: beacon-chain/core/blocks/block_operations.go (hunk 5 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: same question as above

File: beacon-chain/core/helpers/validators.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/validators.go]: maybe i will add error to the function?

File: beacon-chain/core/validators/validator.go (hunk 7 above)
Comment (shayzluf) [reply beacon-chain/core/validators/validator.go]: changed,
thanks

File: beacon-chain/state/getters.go (hunk 8 above)
Comment (shayzluf) [reply beacon-chain/state/getters.go]: should i add error also here?

File: beacon-chain/core/blocks/block_operations.go (hunk 4 above)
Comment (terencechain) [reply beacon-chain/core/blocks/block_operations.go]: so a bad deposit can fail a whole block? i always thought deposit is an independent operation

File: beacon-chain/core/helpers/validators.go (hunk 6 above)
Comment (terencechain) [reply beacon-chain/core/helpers/validators.go]: i think that's better than 0

File: beacon-chain/core/blocks/block_operations.go (hunk 4 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: ok,
changing to continue

File: beacon-chain/core/blocks/block_operations.go (hunk 5 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: changed to continue

Comment (nisdas) [reply beacon-chain/core/blocks/block_operations.go]: why continue ? it should fail then

File: beacon-chain/core/blocks/block_operations.go [hunk 9]
```diff
@@ -830,6 +867,9 @@ func ProcessDeposits(
 	var err error
 	deposits := body.Deposits
 	for _, deposit := range deposits {
+		if deposit == nil || deposit.Data == nil {
+			continue
```
Comment (nisdas) [inline beacon-chain/core/blocks/block_operations.go]: again it should fail over here

File: shared/bls/bls.go [hunk 10]
```diff
@@ -269,6 +269,9 @@ func (s *Signature) Marshal() []byte {
 //    fork_version = state.fork.previous_version if epoch < state.fork.epoch else state.fork.current_version
 //    return compute_domain(domain_type, fork_version)
 func Domain(domainType []byte, forkVersion []byte) uint64 {
+	if len(domainType) < 4 || len(forkVersion) < 4 {
+		return 0
```
Comment (nisdas) [inline shared/bls/bls.go]: 0 is a valid domain

File: beacon-chain/core/helpers/validators.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/validators.go]: added error here, should i do the same for ActivationEligibilityEpoch? quite a lot of changes

File: beacon-chain/core/blocks/block_operations.go (hunk 9 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: i changed to continue after a discussion with @terencechain i also thought this is a cause for failure. no nil deposit should be in a list of deposits. changing back to failure

File: beacon-chain/core/blocks/block_operations.go [hunk 11]
```diff
@@ -1024,6 +1082,9 @@ func ProcessVoluntaryExits(
 ) (*stateTrie.BeaconState, error) {
 	exits := body.VoluntaryExits
 	for idx, exit := range exits {
+		if exit == nil || exit.Exit == nil {
+			continue
```
Comment (shayzluf) [inline beacon-chain/core/blocks/block_operations.go]: @nisdas @terencechain i think it should fail. why should a block contain nil entries?

File: beacon-chain/core/blocks/block_operations.go [hunk 12]
```diff
@@ -394,6 +412,9 @@ func ProcessProposerSlashings(
 ) (*stateTrie.BeaconState, error) {
 	var err error
 	for idx, slashing := range body.ProposerSlashings {
+		if slashing == nil {
+			continue
```
Comment (shayzluf) [inline beacon-chain/core/blocks/block_operations.go]: @nisdas @terencechain i think it should fail. why should a block contain nil entries?

File: shared/bls/bls.go (hunk 10 above)
Comment (shayzluf) [reply shared/bls/bls.go]: adde error to domain function

File: beacon-chain/core/blocks/block_operations.go (hunk 9 above)
Comment (terencechain) [reply beacon-chain/core/blocks/block_operations.go]: I see. I think i agree fail is more sensible here. I chose `cont` at first to be more align with the spec logic. Switching to fail is good with me

File: beacon-chain/core/blocks/block_operations.go (hunk 12 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: changed to fail

File: beacon-chain/core/blocks/block_operations.go (hunk 11 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: changed to fail

File: beacon-chain/core/blocks/block_operations.go (hunk 9 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: thanks

File: beacon-chain/state/getters.go (hunk 8 above)
Comment (shayzluf) [reply beacon-chain/state/getters.go]: returned FarFutureEpoch

File: beacon-chain/core/helpers/validators.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/validators.go]: in ActivationEligibilityEpoch i returned far future epoch

File: beacon-chain/state/getters.go [hunk 13]
```diff
@@ -13,36 +15,54 @@ import (
...
+	}
 	return v.validator.EffectiveBalance
 }
 
 // ActivationEligibilityEpoch returns the activation eligibility epoch of the
 // read only validator.
 func (v *ReadOnlyValidator) ActivationEligibilityEpoch() uint64 {
+	if v == nil || v.validator == nil {
```
Comment (nisdas) [inline beacon-chain/state/getters.go]: i am not sure this is correct, we shouldnt be returning zero values here for `readOnlyValidators`. The correct way is to to check for nil validators when we create the `ReadOnlyValidator` object.

File: shared/bls/bls.go (hunk 10 above)
Comment (nisdas) [reply shared/bls/bls.go]: ?

File: beacon-chain/state/getters.go (hunk 13 above)
Comment (rauljordan) [reply beacon-chain/state/getters.go]: Agree, 0 value for uint64s are very dangerous if we aren't careful @shayzluf

Comment (shayzluf) [reply beacon-chain/state/getters.go]: it is not 0 value it FarFutureEpoch

Comment (nisdas) [reply beacon-chain/state/getters.go]: The issue still stands, you are returning a value for a nil validator. `ReadOnlyValidator` is only used in 1 place, so 
we should still be able to check over there, if any nil validators exist and return an error there.

Comment (shayzluf) [reply beacon-chain/state/getters.go]: its far future epoch and not 0 that is being returned

Comment (shayzluf) [reply beacon-chain/state/getters.go]: removed it and handled it in ReadOnlyValidator creation

File: shared/bls/bls.go (hunk 10 above)
Comment (shayzluf) [reply shared/bls/bls.go]: change to [4]bytes

**Pull request merged**